### ISSUE TITLE
BatchNormalization calculates batch mean and variance even in inference mode

### ISSUE THREAD

**Issue by physicsrob** [keras-team/keras#9455]
I'm not completely confident in the issue title, but I find it plausible and would love it if someone from the keras community helped verify/fix the issue.

Here's the steps to reproduce:
1. Build any model with one or more batch norm layers.  In my case it was a functional API model, although I doubt that is necessary to reproduce.
2. run model.predict() (wrap it in a timer of some sort)
3. At normalization.py:174 (as of master) add `training=True`
4. run model.predict()
5. observe that the model runs significantly faster.  For my model I get about a 3X speed increase.

**Comment by physicsrob** [comment]
Investigating a bit more, I think I understand what's going on.  At the end of `BatchNormalization.call` we return a `K.in_training_phase` which basically does a `tf.cond()` on the training phase to use either the training mode batch norm or the inference batch norm.  Tensorflow executed both branches, so even though the calculation is correct, we end up calculating the mean and variance even in inference mode.

Does that make sense?  Any thoughts on a solution?

**Comment by Dref360** [comment]
This is as designed (by TF). So I don't think we can avoid it.
Ref : https://github.com/tensorflow/tensorflow/issues/10800

**Comment by physicsrob** [comment]
I think I agree that it can't be avoided from a tensorflow perspective if tf.cond is used (although I'm not an expert on TF graphs, so who knows?)  But regardless, there's already logic in BatchNorm to short circuit the moments calculation if you pass training=False as an argument.  That's pretty awkward, however.  In my case I end up doing something along the lines of ```
prev = BatchNormalization()(prev, training=training)
```
where training is a global I manage.  It seems like there's got to be a better way, and if there's not, at the very least I would expect the documentation to be modified to warn the users about the overhead in inference mode and the way to avoid it.
